Add bookings

DenisEdikh opened the pull request:
added functionality according to the terms of reference

DenisEdikh committed:
delete because it is not used

DenisEdikh committed:
added properties of DB, time_zone and logging

DenisEdikh committed:
added new functional by technical specification 15

DenisEdikh committed:
added new method invalidRequest

DenisEdikh committed:
added annotation of Hibernate

DenisEdikh committed:
added new mapping methods

DenisEdikh committed:
deleted unused method

DenisEdikh committed:
changed interface (extends JpaRepository, QuerydslPredicateExecutor)

DenisEdikh committed:
added new methods

DenisEdikh committed:
added new method "createComment", changed "userId" to "ownerId", changed returned type in the "getItemById" method from Item to ItemDto

DenisEdikh committed:
1. added new method "createComment", "getCommentsByItemId", "checkItemById", "checkOwnerByItem", "checkEndTime".
2. changed method "getById" (changed returned type, added new fields to Dto)

DenisEdikh committed:
added new dependencies

DenisEdikh committed:
temporarily deleted so that the tests would pass

DenisEdikh committed:
checkstyle

DenisEdikh committed:
1. created separate method to get the user from the database
2. corrected method "checkEmail"
3. changed create to save in method "create"
4. changed delete to deleteById in method "delete"

DenisEdikh committed:
change branch name

kesch9 commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,49 @@
+    end_time   timestamp NOT NULL,
+    item_id    bigint    NOT NULL,
+    user_id    bigint    NOT NULL,
+    status     int       NOT NULL default 2,
```
Обязательно
Статус должен храниться в varchar, а в поле модели указывается
@Enumerated(EnumType.STRING)
private Status status
Конверторы не нужны будут

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,49 @@
+CREATE TABLE IF NOT EXISTS comments
+(
+    id           bigint                      NOT NULL GENERATED ALWAYS AS IDENTITY,
+    text         text                        NOT NULL,
```
Можно лучше
Лучше не использовать в названиях зарезервированные слова, text это тип в БД

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -1,7 +1,69 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+
```
Можно лучше
Лишний пробел между аннотациями

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -1,7 +1,69 @@
+    @JoinColumn(name = "user_id")
+    private User booker;
+    @Convert(converter = StatusConverter.class)
+    private Status status = Status.WAITING;
```
Здесь надо будет писать
    @Enumerated(EnumType.STRING)
    private Status status = Status.WAITING;

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -1,7 +1,69 @@
+    private Status status = Status.WAITING;
+
+    @Override
+    public final boolean equals(Object o) {
```
Можно и не переопределять equals и hashcode у моделей

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -1,7 +1,69 @@
+        if (this == o) return true;
+        if (o == null) return false;
+        Class<?> oEffectiveClass = o instanceof HibernateProxy
+                ? ((HibernateProxy) o).getHibernateLazyInitializer().getPersistentClass()
```
Это лишнее вызывать proxy для определения equals и hashcode

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/booking/BookingServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+        LocalDateTime ldt = LocalDateTime.now();
+        return switch (state) {
+            case ALL -> BookingMapper.toBookingDto(bookingRepository
+                    .findBookingByBookerIdOrderByStartDesc(bookerId));
```
Обязательно
В метода репозиторий, вместо явного указания сортировки используйте Sort, если вам будет другой способ сортировки придется создавать новый метод, а так просто поменяется Sort.
Создается вот так
Sort newestFirst = Sort.by(Sort.Direction.DESC, "start");
а используется просто добавляя в метод его
findBookingByBookerId(bookerId, sort)

DenisEdikh replied:
Откорректировано.

kesch9 commented on src/main/java/ru/practicum/shareit/booking/BookingServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    @Override
+    public BookingDto getById(Long bookingId, Long userId) {
```
Обязательно
Еще хорошо добавить проверку, есть ли вообще такой пользователь или нет

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/booking/BookingServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+        if (!item.getOwner().getId().equals(userId)) {
+            log.warn("Only owner can change status");
+            throw new InvalidRequestException("Only owner can change status");
+        }
```
Обязательно
Должна быть еще проверка, что
booking.getStatus() != Status.WAITING
в этом случае кидаем ошибку

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/item/Comment.java:
```diff
@@ -0,0 +1,63 @@
+@AllArgsConstructor
+@Entity
+@Table(name = "comments")
+public class Comment {
```
Обязательно
У комментариев должен быть свой пакет comment, нужно добавить этот пакет и перенести туда все сущности относящиеся к комментариям

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/user/UserServiceImpl.java:
```diff
@@ -15,43 +16,42 @@
+    @Transactional
     public UserDto create(NewUserDto newUserDto) {
         log.debug("Started checking email user in method create");
+        checkEmail(newUserDto.getEmail());
```
Обязательно
Проверка по email уже не нужна, проверка будет на уровне БД

DenisEdikh replied:
Откорректировано

kesch9 commented on src/main/java/ru/practicum/shareit/booking/BookingServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+public class BookingServiceImpl implements BookingService {
+    private final BookingRepository bookingRepository;
+    private final UserService userService;
+    private final ItemService itemService;
```
Можно лучше
Вместо сервисных классов, лучше работать с репозиториями в сервисных классах

DenisEdikh replied:
Откорректировано во всех сервисах

kesch9 commented on src/main/java/ru/practicum/shareit/user/UserServiceImpl.java:
```diff
@@ -61,15 +61,18 @@ public List<UserDto> getAll() {
+        return UserMapper.toUserDto(checkUserById(userId));
+    }
+
+    public User checkUserById(Long userId) {
```
Обязательно
Нет @Override над методом

DenisEdikh replied:
Откорректировано.
Метод убран в приватные, из интерфейса убран

kesch9 reviewed (commented):
Нужно немного доработать код

DenisEdikh committed:
deleted overriding equals and hashcode method

DenisEdikh committed:
deleted as unused

DenisEdikh committed:
changed package

DenisEdikh committed:
changed package and added @Column for field "text"

DenisEdikh committed:
deleted as unused

DenisEdikh committed:
added to methods argument sort and changed names of methods

DenisEdikh committed:
deleted constructor, field and number of enum

DenisEdikh committed:
1. changed name from "text" to "description" in table "comments"
2. changed type of field "status" in table bookings

DenisEdikh committed:
1. changed services to repositories
2. corrected methods of booking repositories in methods "getAllByOwner","getAllByState"
3. added new method "checkUserIsContained"

DenisEdikh committed:
1. changed package of comment in import

DenisEdikh committed:
changed package of comment in import

DenisEdikh committed:
deleted unused method "checkUserById"

DenisEdikh committed:
deleted unused method "checkEmail", because now verification is carried out using a database

DenisEdikh committed:
1. added new method "checkUserIsContained"
2. change user service to user repositories

DenisEdikh committed:
deleted unused method "checkItemById"

kesch9 reviewed (approved):
Approve
